Ilariae/solana shims (#532)
wormhole-foundation/wormhole-docs: products/messaging/guides/solana-shims/sol-verification.md

Reviewer: In my opinion, this article should focus more concretely on 1) a new integrator who needs to verify VAAs on Solana and 2) an existing integrator who wants to migrate to using the Verification Shim. There is an example program instruction [here](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/programs/wormhole-integrator-example/src/instructions/consume_vaa.rs).

description: Efficiently verify Wormhole VAAs on Solana using the Verification Shim, which avoids persistent rent-exempt accounts while keeping full security guarantees.
categories: Basics
---

# Solana VAA Verification via Shim

Verifying VAAs on Solana with the legacy core bridge requires creating multiple rent-exempt accounts for signatures and posted VAAs. These accounts persist even after verification is complete, which increases costs and bloats the on-chain state.

The verification shim solves this by replacing the core bridge verification flow with its own instructions:

- `post_signatures`: Accumulates Guardian signatures into a temporary account.
- `verify_hash`: Validates the VAA by checking the signatures against the active Guardian set and ensuring quorum.
- `close_signatures`: Closes the temporary account to reclaim lamports.

Because the shim avoids leaving permanent accounts behind, verification becomes much cheaper while keeping the same security guarantees.

This page introduces the Verification Shim, explains how it works, and shows how integrators can adopt it in place of the core bridge’s `verify_signatures` and `post_vaa` functions.

> For more background, see [Verification Shim concept section](/docs/products/messaging/concepts/solana-shim/#verification-shim){target=\_blank}. 

Reviewer 2: ```suggestion
For more background, see the [Verification Shim concept section](/docs/products/messaging/concepts/solana-shim/#verification-shim){target=\_blank}. 
```

## Prerequisites

To interact with the verification shim, you'll need the following:

> - [Rust and Solana CLI](https://docs.solana.com/cli/install-solana-cli-tools){target=\_blank} installed.  

Reviewer 2: ```suggestion
- [Rust and Solana CLI installed](https://docs.solana.com/cli/install-solana-cli-tools){target=\_blank}.  
```

> - [Anchor](https://www.anchor-lang.com/docs/installation){target=\_blank}

Reviewer 2: The other bullets in this list have some form of full sentence or explanation following the subject can one be add here for Anchor, so that it can match better with a period?

> - [Anchor installed](https://www.anchor-lang.com/docs/installation){target=\_blank}

Reviewer 2: ```suggestion
- [Anchor installed](https://www.anchor-lang.com/docs/installation){target=\_blank}.
```

```

## Accounts

You’ll wire three accounts for verification:

- `guardian_set`: Core bridge `GuardianSet` PDA for the VAA’s `guardianSetIndex` (shim checks derivation).
- `guardian_signatures`: Temporary account created by `post_signatures` (shim checks ownership & discriminator).
- `wormhole_verify_vaa_shim`: The verification shim program.

```rs
--8<-- 'code/products/messaging/guides/shims/consume_vaa.rs:10:21'
```

Here, `guardian_set` is a core bridge PDA, and `guardian_signatures` is created and owned by the verification shim. Derive `guardian_set = PDA(["GuardianSet", index_be_bytes], CORE_BRIDGE_PROGRAM_ID)` using the `guardianSetIndex` from the VAA header (big-endian), compute its bump, and pass that bump into your instruction.

## Verify the VAA

The `consume_vaa` function computes the digest, calls the shim’s `verify_hash`, and then proceeds with your logic. This step validates Guardian signatures and quorum against the active Guardian set for the VAA’s `guardianSetIndex`, then lets your program proceed without persisting a `PostedVAA`. 

```rs
--8<-- 'code/products/messaging/guides/shims/consume_vaa.rs:23'
```

## Limitations and Security Considerations

> - You must be the payer and/or account owner to reclaim lamports from `GuardianSignatures` account.

Reviewer 2: ```suggestion
- You must be the payer and/or account owner to reclaim lamports from the `GuardianSignatures` account.
```

> - The verification proof is not permanent on-chain record unless you keep the account.

Reviewer 2: ```suggestion
- The verification proof is not a permanent on-chain record unless you keep the account.
```

> - Compute usage (CU) is higher for the rent-efficient pattern, but total cost is dramatically lower than keeping permanent accounts.

Reviewer 2: ```suggestion
- Compute usage (CU) is higher for the rent-efficient pattern, but the total cost is dramatically lower than keeping permanent accounts.
```

---
title: Efficient VAA Verification on Solana (Shim)
description: Efficiently verify Wormhole VAAs on Solana without leaving rent-exempt accounts, using the core bridge’s standard instructions.
categories: Basics
---

# Efficient VAA Verification on Solana (Shim)

This guide explains how to efficiently verify Wormhole VAAs on Solana by leveraging the core bridge’s [`verify_signatures`](https://github.com/wormhole-foundation/wormhole/blob/main/solana/bridge/program/src/api/verify_signature.rs){target=\_blank} and [`post_vaa`](https://github.com/wormhole-foundation/wormhole/blob/main/solana/bridge/program/src/api/post_vaa.rs){target=\_blank} instructions, and cleaning up temporary accounts after use.

The goal is to accumulate all guardian signatures into a temporary `SignatureSet` account using `verify_signatures`, verify the VAA and guardian set using `post_vaa`, and immediately close any accounts you created for this process.

> Note that the shim is not a new contract or program. Instead, it’s a rent-efficient usage pattern of the existing Wormhole Core Bridge instructions (`verify_signatures` and `post_vaa`). By following this approach, you only create temporary accounts for the verification process and clean them up immediately afterward—leaving no permanent on-chain state.

Reviewer 3: ```suggestion
Note that the shim is not a new contract or program. Instead, it’s a rent-efficient usage pattern of the existing Wormhole Core Bridge instructions (`verify_signatures` and `post_vaa`). By following this approach, you only create temporary accounts for the verification process and clean them up immediately afterward, leaving no permanent on-chain state.
```

Reviewer 3: Aren't shims lightweight programs? Here you're saying that they are not programs.

Reviewer 3: ```suggestion
Note that the shim is not a new program. Instead, it’s a rent-efficient usage pattern of the existing Wormhole Core Bridge instructions (`verify_signatures` and `post_vaa`). By following this approach, you only create temporary accounts for the verification process and clean them up immediately afterward, leaving no permanent on-chain state.
```
Perhaps we shouldn't use the term 'contract' here, as Solana doesn't use contracts.

Author: From my understanding, the emission shim is an actual program you deploy, while the verification shim is not a new program but more of an approach that uses the core bridge instructions

For more background, see [Solana Shims concept page](/docs/products/messaging/concepts/solana-shim/){target=\_blank}. For deployment steps, see [Shim Deployment guide](/docs/products/messaging/guides/solana-shims/shim-deployment/){target=\_blank}.

## How It Works

The verification shim replaces the legacy multi-account pattern with a flow where you only create a temporary signature set account. After verification, you can close it to reclaim your lamports.

> 1. **Create a temporary SignatureSet account**: Fund it as rent-exempt for the required size.

Reviewer 3: ```suggestion
1. **Create a temporary `SignatureSet` account**: Fund it as rent-exempt for the required size.
```

let digest = keccak::hash(message_hash.as_slice()).to_bytes();
```

A QueryResponse digest can be computed as follows:

```rust
use wormhole_query_sdk::MESSAGE_PREFIX;
let message_hash = [
  MESSAGE_PREFIX,
  &solana_program::keccak::hashv(&[&bytes]).to_bytes(),
].concat();
let digest = keccak::hash(message_hash.as_slice()).to_bytes();
```

## Caveats

This shim will need to include the following patch made to the core bridge when calculating the expiry for guardian sets.

```rust
pub fn is_active(&self, timestamp: &u32) -> bool {
    // Note: This is a fix for Wormhole on mainnet.  The initial guardian set was never expired
    // so we block it here.
    if self.index == 0 && self.creation_time == 1628099186 {
        false
    } else {
        self.expiration_time == 0 || self.expiration_time >= *timestamp
    }
}
```

Unlike the core bridge, it will not need to perform signature set deny-listing, as it does the verification directly through `secp256k1_recover`. 

Since it is planned to be non-upgradeable, any similar mitigation strategies will not be possible. e.g. the only way to expire a guardian set will be for the core bridge to properly expire it.

## Deployment

> No special deployment is required—these are standard core bridge instructions. For advanced multi-program flows, see the [Solana Shim Deployment Guide](/docs/products/messaging/guides/solana-shims/shim-deployment/){target=\_blank} for detailed deployment steps.

Reviewer 3: ```suggestion
No special deployment is required; these are standard core bridge instructions. For advanced multi-program flows, see the [Solana Shim Deployment Guide](/docs/products/messaging/guides/solana-shims/shim-deployment/){target=\_blank} for detailed deployment steps.
```

---
title: Efficient VAA Verification on Solana (Shim)
description: Efficiently verify Wormhole VAAs on Solana without leaving rent-exempt accounts, using the core bridge’s standard instructions.
categories: Basics
---

> # Efficient VAA Verification on Solana (Shim)

Reviewer: This section is incorrect. The goal is to use `post_signatures`, `verify_hash`, and `close_signatures` on the Verification Shim. No calls to the Core Bridge are needed, as this replaces them.

This guide explains how to efficiently verify Wormhole VAAs on Solana by leveraging the core bridge’s [`verify_signatures`](https://github.com/wormhole-foundation/wormhole/blob/main/solana/bridge/program/src/api/verify_signature.rs){target=\_blank} and [`post_vaa`](https://github.com/wormhole-foundation/wormhole/blob/main/solana/bridge/program/src/api/post_vaa.rs){target=\_blank} instructions, and cleaning up temporary accounts after use.

The goal is to accumulate all guardian signatures into a temporary `SignatureSet` account using `verify_signatures`, verify the VAA and guardian set using `post_vaa`, and immediately close any accounts you created for this process.

> !!!info "What does 'shim' mean here?"

Reviewer: see prior comment on how this is inaccurate.

    - For emission, the shim is a new lightweight program.  
    - For verification, the 'shim' is just a smart use of existing instructions.
    - See [Shim Emission and Verification](/docs/products/messaging/concepts/solana-shim/#shim-emission-and-verification){target=\_blank} for more details.

For more background, see [Solana Shims concept page](/docs/products/messaging/concepts/solana-shim/){target=\_blank}. For deployment steps, see [Shim Deployment guide](/docs/products/messaging/guides/solana-shims/shim-deployment/){target=\_blank}.

## How It Works

The verification shim replaces the legacy multi-account pattern with a flow where you only create a temporary signature set account. After verification, you can close it to reclaim your lamports.

> 1. **Create a temporary `SignatureSet` account**: Fund it as rent-exempt for the required size.

Reviewer: The verification shim [IDL](https://github.com/wormhole-foundation/wormhole/blob/main/svm/wormhole-core-shims/anchor/idls/wormhole_verify_vaa_shim.json) refers to this as a `GuardianSignatures` account and handles the creation and funding of it.

> 2. **Call `verify_signatures`** as many times as needed, using the secp256k1 syscall and all guardian signatures. The SignatureSet account will accumulate valid signatures.

Reviewer: no call to `verify_signatures` is used. the **unverified** signatures are posted to a temporary account using `post_signatures`

> 3. **Call `post_vaa`** to check guardian set validity, consensus, and VAA integrity.

Reviewer: call `verify_hash` from within your program.

   - If verification succeeds, proceed with your on-chain logic (e.g., updating state, processing transfers).

>    - The `PostedVAA` account is created, but you do **not** need to keep it around for rent; it can be closed after use if you control it.

Reviewer: there is no `PostedVAA` account.

> 4. **Immediately close** the `SignatureSet` and any `PostedVAA` accounts you created to reclaim lamports, if you are the payer.

Reviewer: Immediately close the `GuardianSignatures` account via `close_signatures`

```mermaid
graph LR
    A[Create SignatureSet] --> B[verify_signatures]
    B --> C[post_vaa]
    C --> D[Process Logic]
    D --> E[Close SignatureSet & PostedVAA]
```

## Verify VAA

This instruction is intended to be invoked via CPI call. It verifies a digest against a GuardianSignatures account and a core bridge GuardianSet. Prior to this call, and likely in a separate transaction, `post_signatures` must be called to create the account. Immediately after this call, `close_signatures` should be called to reclaim the lamports.

A v1 VAA digest can be computed as follows:

```rust
let message_hash = &solana_program::keccak::hashv(&[&vaa_body]).to_bytes();
let digest = keccak::hash(message_hash.as_slice()).to_bytes();
```

A QueryResponse digest can be computed as follows:

```rust
use wormhole_query_sdk::MESSAGE_PREFIX;
let message_hash = [
  MESSAGE_PREFIX,
  &solana_program::keccak::hashv(&[&bytes]).to_bytes(),
].concat();
let digest = keccak::hash(message_hash.as_slice()).to_bytes();
```

> ## Caveats

Reviewer: This caveats section was only relevant to the development of the verification shim and does not need to be reproduced here.

This shim will need to include the following patch made to the core bridge when calculating the expiry for guardian sets.

```rust
pub fn is_active(&self, timestamp: &u32) -> bool {
    // Note: This is a fix for Wormhole on mainnet.  The initial guardian set was never expired
    // so we block it here.
    if self.index == 0 && self.creation_time == 1628099186 {
        false
    } else {
        self.expiration_time == 0 || self.expiration_time >= *timestamp
    }
}
```

Unlike the core bridge, it will not need to perform signature set deny-listing, as it does the verification directly through `secp256k1_recover`. 

Since it is planned to be non-upgradeable, any similar mitigation strategies will not be possible. e.g. the only way to expire a guardian set will be for the core bridge to properly expire it.

## Deployment

> No special deployment is required; these are standard core bridge instructions. For advanced multi-program flows, see the [Solana Shim Deployment Guide](/docs/products/messaging/guides/solana-shims/shim-deployment/){target=\_blank} for detailed deployment steps.

Reviewer: They are not Core Bridge instructions, they are Verification Shim instructions. 

```
For advanced multi-program flows, see the [Solana Shim Deployment Guide](/docs/products/messaging/guides/solana-shims/shim-deployment/){target=\_blank} for detailed deployment steps.
```

This statement does not make sense, please remove it.

## Limitations and Security Considerations

> - You must be the payer and/or account owner to reclaim lamports from SignatureSet and PostedVAA accounts.

Reviewer: ```suggestion
- You must be the payer and/or account owner to reclaim lamports from GuardianSignatures account.
```

- The verification proof is ephemeral—no permanent on-chain record unless you keep the account.
- Compute usage (CU) is higher for the rent-efficient pattern, but total cost is dramatically lower than keeping permanent accounts.
- All validation guarantees remain as strong as with the legacy method.
- If you do not close accounts you create, rent will be lost as before.

> - Only close accounts after all logic is complete; never close in the middle of validation.

Reviewer: I'm not sure what this means. The integrator cannot "close in the middle of validation"
